### PR TITLE
Allow navigation between blocks with arrow keys

### PR DESCRIPTION
**Pull request by ellatrix**
This PR adds arrow key navigation between blocks in a very basic way, just like tabs would work. I took the approach of looking at focusable nodes instead of passing props, so block implementors don't have to worry about this as well.

### CODE REVIEW

**Comment by youknowriad** [comment]
Nice! 👍 I'd suggest adding the same behavior to `VisualEditorBlock` if the focused element is the block container node, this would allow navigating from top to bottom without "stop".

Also, I noticed a "bug" in the button block where the focus moves to the link Editable. We could add tabindex="-1" maybe.

Regardless, this is great 👍

File: blocks/editable/index.js [hunk 1]
```diff
@@ -194,10 +194,41 @@ export default class Editable extends wp.element.Component {
 	}
 
 	onKeyDown( event ) {
+		const { keyCode } = event;
+		const moveUp = ( keyCode === UP || keyCode === LEFT ) && this.isStartOfEditor();
+		const moveDown = ( keyCode === DOWN || keyCode === RIGHT ) && this.isEndOfEditor();
+		const selectors = [
+			'*[contenteditable="true"]',
+			'*[tabindex]',
```
Comment (youknowriad) [inline blocks/editable/index.js]: should we exclude tabindex=-1?

Comment (aduth) [inline blocks/editable/index.js]: Yeah, this won't accommodate all possibilities. Another being `contenteditable` without an explicitly `true` value (`contenteditable`, `contenteditable=""`). There's also `select`, `button`, `object`, `area`, `a`, though dunno if we want to test all of them in this behavior.

I'm forgetting the resource I used in creating it, but this implementation is close to spec for "focusable":

https://codepen.io/aduth/pen/wdRzxM

Maybe also worth moving out as a separate utility.

Comment (youknowriad) [inline blocks/editable/index.js]: Just found this library that looks interesting https://github.com/medialize/ally.js/blob/master/docs/api/README.md And it has a utility to find the tabbables elements https://github.com/medialize/ally.js/blob/master/docs/api/query/tabbable.md

Comment (aduth) [inline blocks/editable/index.js]: Wow, proper tabbable detection is complex!

https://github.com/medialize/ally.js/blob/master/src/is/tabbable.js

Comment (iseulde) [inline blocks/editable/index.js]: Wow, yeah, that's a lot. I'm all for using a library for this if that's better.

**Comment by iseulde** [comment]
@youknowriad Trying something slightly different at a block level... This approach looks at the range. If the range is the same, this means that nothing happened when pressing the arrow keys, which either means that the caret is at the start or end of text input, or that it's not text input. This seems to be good because it will handle more than just `Editable`.

**Comment by iseulde** [comment]
To do: use that library.

**Comment by youknowriad** [comment]
This seems to work great for all our current blocks, but I wonder if it's not too aggressive. I mean, maybe a block with an input or something don't want to lose focus if we accidentally click on arrows. This question is I think hard to answer right now, so I'm ok with moving forward, I haven't seen any glitch (aside the link boundaries)

**Comment by iseulde** [comment]
Not sure why the tests don't pass.

Yeah, the links boundaries are an issue with either approach.

**Comment by notnownikki** [comment]
Purely from a user's perspective, this is so much better.

I'm undecided on how it handles the form fields... On one hand, it's not standard behaviour and yeah, we might lose focus accidentally. But on the other, it's _great_ if I'm going from the bottom of the document to the top by holding down the 'up' arrow. I can completely navigate my post without leaving the arrow keys, and I only tab if I want to get into toolbars and other editing controls. However, if I was going up a document and ended up in an input field and arrow keys no longer took me up, that's expected behaviour and tabbing is the normal keyboard way of getting out of that, and would do what I expect.

Nice!

**Comment by iseulde** [comment]
Working on redoing this with props right now.

**Comment by iseulde** [comment]
I think I'll leave table alone for now and do that in a separate PR. You can still navigate over it though.

**Comment by youknowriad** [comment]
I like the flexibility of the current path because the block author has all the powers but I think I prefer the approach relying on "Tabbable" elements because it removes some burden from the block author. I hope I didn't push you to the wrong direction. Happy to follow the majority here. 

I wonder if there's not a middleground where we can do tabbables but we can deactivate it for more control and rely on props. Anyway, I think we should start with tabbables only. What do you think?

**Comment by iseulde** [comment]
Oh no, I really wish I saved the previous approach and didn't force push 🙈

**Comment by iseulde** [comment]
@youknowriad So I really don't kind either approach, I can open a new PR with the previous approach. I'm not sure what the middle ground is you're suggesting. :) Maybe some more eyes could help here? @matias What do you think?

**Comment by mtias** [comment]
Fixes #1723.

**Comment by iseulde** [comment]
Follow up PR: #1778.

**Comment by aduth** [comment]
Should this be closed as superseded by #1778 ?

**Comment by iseulde** [comment]
Yep

**Pull request closed without merging**